src/complex/DataStructure/AbstractDataStore.hpp

```diff
@@ -44,9 +44,13 @@ class AbstractDataStore : public IDataStore
     using pointer = T*;
     using reference = T&;
 
-    Iterator() = delete;
+    Iterator()
+    : m_DataStore(nullptr)
+    , m_Index(0)
+    {
+    }
 
-    Iterator(AbstractDataStore& dataStore, usize index)
+    Iterator(AbstractDataStore* dataStore, usize index)
```

Comment:
```suggestion
    Iterator(AbstractDataStore& dataStore, usize index)
```
We can still use a reference in the signature of the constructor even if we store it as a pointer. I also think we should leave a comment that the iterator needs to be default constructible for some standard library algorithm implementations. And this change could be its own PR.